src/elisp/treemacs-workspaces.el

```diff
@@ -138,18 +138,26 @@ Will return `point-min' if there is no next project."
   (inline-quote
    (setq treemacs--project-positions (make-hash-table :test #'equal :size 20))))
 
+(define-inline treemacs-project->key (project)
+  "Get the hash table key of PROJECT."
+  (inline-letevals (project)
+    (inline-quote
+     (if (symbolp ,project)
```

Comment:
This happens with root-level extensions [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L443) and [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L465) - it might be considered a bug though...

Reply:
I changed this to `treemacs-project-p`, since `:root-key-form` does not have to be a symbol.

Reply:
Not a bug. This is an explicitly wanted exception. Using a project object as a key to save its position works most of the time, but variadic extensions really throw a wrench into those gears. If you don't know how many top-level extension nodes you produce and what they will be called then insisting on consistenly using structs as keys is more trouble than it's worth and using the project's key in such cases is IMO the lesser evil.

Reply:
Thanks for the explanation, makes sense (mostly - I'm not super familiar with variadic extensions though).

Reply:
Neither am I - they are the next big thing I will be tackling. A good use-case is a package like https://github.com/emacs-lsp/lsp-treemacs. With variadic extensions it will be able to display any combination of 0..3 top-level nodes: Infos and/or Warnings and/or Errors.

Updating will be the interesting part, since entire subtrees may disappear or be added to the dom. There's a lengthy discussion in https://github.com/Alexander-Miller/treemacs/issues/331 if you're interested.

Reply:
I'm definitely interested, as I want to update the nodes in `treemacs-bookmark` when bookmarks change.

Reply:
There are already methods to update and delete single nodes, see `treemacs-update-node` and `treemacs-delete-single-node`. You only need to wait if your extension is variadic (I haven't tried your bookmark extension yet, it's next on my todo list after integrating this).